pkg/template/servicebroker/servicebroker.go

```diff
@@ -1,62 +1,68 @@
 package servicebroker
 
 import (
-	"k8s.io/apiserver/pkg/authentication/user"
-	restclient "k8s.io/client-go/rest"
-	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
-	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"errors"
+	"time"
 
 	authclient "github.com/openshift/origin/pkg/auth/client"
 	"github.com/openshift/origin/pkg/client"
-	"github.com/openshift/origin/pkg/client/cache"
-	"github.com/openshift/origin/pkg/controller/shared"
+	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+	templateapi "github.com/openshift/origin/pkg/template/api"
+	templateinformer "github.com/openshift/origin/pkg/template/generated/informers/internalversion/template/internalversion"
 	templateclientset "github.com/openshift/origin/pkg/template/generated/internalclientset"
 	internalversiontemplate "github.com/openshift/origin/pkg/template/generated/internalclientset/typed/template/internalversion"
+	templatelister "github.com/openshift/origin/pkg/template/generated/listers/template/internalversion"
+	"k8s.io/apiserver/pkg/authentication/user"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 )
 
 type Broker struct {
-	secretsGetter      internalversion.SecretsGetter
-	localSAR           client.LocalSubjectAccessReviewsNamespacer
+	oc                 *client.Client
+	kc                 kclientset.Interface
 	templateclient     internalversiontemplate.TemplateInterface
 	restconfig         restclient.Config
-	lister             cache.StoreToTemplateLister
+	informer           cache.SharedIndexInformer
+	lister             templatelister.TemplateLister
 	templateNamespaces map[string]struct{}
 }
 
-func NewBroker(restconfig restclient.Config, localSAR client.LocalSubjectAccessReviewsNamespacer, secretsGetter internalversion.SecretsGetter, informers shared.InformerFactory, namespaces []string) *Broker {
+func NewBroker(restconfig restclient.Config, informer templateinformer.TemplateInformer, namespaces []string) *Broker {
+	informer.Informer().AddIndexers(cache.Indexers{
+		templateapi.TemplateUIDIndex: func(obj interface{}) ([]string, error) {
+			return []string{string(obj.(*templateapi.Template).UID)}, nil
+		}})
+
 	templateNamespaces := map[string]struct{}{}
 	for _, namespace := range namespaces {
 		templateNamespaces[namespace] = struct{}{}
 	}
 
+	// TODO: avoid creating a new client here
+	restconfig = authclient.NewImpersonatingConfig(&user.DefaultInfo{Name: "system:serviceaccount:openshift-infra:" + bootstrappolicy.InfraTemplateServiceBrokerServiceAccountName}, restconfig)
```

Comment:
> It would definitely be better not as an impersonating config, but I'm not sure how to create this before the API server is up (or if I should do an async initialisation & check for initialisation on each API call?)

Similar to controllers, you can lookup your SA token using the privileged loopback until you get one.  You can reject "early" requests with a 'try again later" return code.